### PR TITLE
[WFCORE-1108]:  Wildfly 9.0.1 Final fails next day with WFLYDS0032: Failed to list files in directory.

### PR DESCRIPTION
**Pull request by ehsavoie**
Using java.nio.file.Files to get better exceptions thus understand this kind of strange behaviour.

Jira: https://issues.jboss.org/browse/WFCORE-1108

### CODE REVIEW

**Comment by bstansberry** [comment]
I put the 3.x label on this because at this point we can't make changes like this in 2.x without a JBEAP.

File: deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java [hunk 1]
```diff
@@ -870,16 +881,14 @@ private void scanDirectory(final File directory, final String relativePath, fina
...
-                                    //we need to make sure that the file was not deleted while
-                                    //the scanner was running
-                                    if (child.exists()) {
-                                        scanContext.incompleteFiles.put(child, new IncompleteDeploymentStatus(child, timestamp));
-                                    }
+                                } else //we need to make sure that the file was not deleted while
+                                //the scanner was running
+                                if (child.exists()) {
+                                    scanContext.incompleteFiles.put(child, new IncompleteDeploymentStatus(child, timestamp));
                                 }
```
Comment (bstansberry) [inline deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java]: This whitespace change is incorrect (contrary to established code style).

File: deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java [hunk 2]
```diff
@@ -904,16 +913,14 @@ private void scanDirectory(final File directory, final String relativePath, fina
...
-                                //we need to make sure that the file was not deleted while
-                                //the scanner was running
-                                if (child.exists()) {
-                                    scanContext.incompleteFiles.put(child, new IncompleteDeploymentStatus(child, timestamp));
-                                }
+                            } else //we need to make sure that the file was not deleted while
+                            //the scanner was running
+                            if (child.exists()) {
+                                scanContext.incompleteFiles.put(child, new IncompleteDeploymentStatus(child, timestamp));
                             }
```
Comment (bstansberry) [inline deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java]: This whitespace change is incorrect (contrary to established code style).

File: deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java [hunk 3]
```diff
@@ -1277,19 +1284,28 @@ private void writeFailedMarker(final File deploymentFile, final String failureDe
...
     }
 
-    private static File[] listDirectoryChildren(File directory, FileFilter filter) {
-        File[] result = directory.listFiles(filter);
-        if (result == null) {
-            throw DeploymentScannerLogger.ROOT_LOGGER.cannotListDirectoryFiles(directory);
+    private static List<File> listDirectoryChildren(File directory, DirectoryStream.Filter<Path> filter) {
+        try (DirectoryStream<Path> stream = Files.newDirectoryStream(directory.toPath(), filter)) {
+            final List<File> result = new ArrayList<>();
+            stream.forEach(entry -> result.add(entry.toFile()));
```
Comment (ctomc) [inline deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java]: if you are changing everything to use Path, do it here as well.

Comment (ehsavoie) [inline deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java]: The goal was to get more informations about the error during the file listing 
Going all the way to nio with its quirks is not for this pr tuo make ;)

**Comment by bstansberry** [comment]
I've put Ready for Merge on this. Please configure your IDE though to not make unrelated whitespace/codestyle changes.

File: deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java (hunk 3 above)
Comment (guofengxu) [reply deployment-scanner/src/main/java/org/jboss/as/server/deployment/scanner/FileSystemDeploymentService.java]: Hi Team,
Did you fix this issue that jboss Failed to list files in directory ？

**Comment by guofengxu** [comment]
Hi Team,
I met this issue in my project meanwhile we use jboss version is "Red Hat JBoss Enterprise  7.0.2 GA"
For this issue , when will be fixed? which jboss version would fix this issue?
In other words, how to avoid this issue through other configuration?

**Pull request merged**